src/transformers/cache_utils.py

```diff
@@ -970,3 +974,156 @@ def get_max_length(self) -> Optional[int]:
         # in theory there is no limit because the sliding window size is fixed
         # no matter how long the sentence is
         return None
+
+    def reset(self):
+        self.key_cache.zero_()
+        self.value_cache.zero_()
+
+
+@dataclass
+class EncoderDecoderCache:
+    """
+    Base, abstract class for all encoder-decoder caches. Can be used to hold combinations of self-attention and
+    cross-attention caches.
+    """
+
+    def __init__(self, self_attention_cache: Cache, cross_attention_cache: Cache):
+        self.self_attention_cache = self_attention_cache
+        self.cross_attention_cache = cross_attention_cache
+
+        self.is_updated = {}
+        for layer_idx in range(len(cross_attention_cache.key_cache)):
+            self.is_updated[layer_idx] = bool(cross_attention_cache.get_seq_length(layer_idx) > 0)
```

Comment:
ouch I hate this

Reply:
Let's keep it to one thread! https://github.com/huggingface/transformers/pull/31166#discussion_r1655015013